userspace/su/jni/su.c

```diff
@@ -1,10 +1,113 @@
+#include <stdio.h>
+#include <string.h>
 #include <unistd.h>
 #include <stdlib.h>
+#include <stdbool.h>
 #include <sys/prctl.h>
 
-int main(){
+#define CMD_GRANT_ROOT 0
+#define CMD_GET_VERSION 2
+
+static void help(int status) {
+
+    printf("KernelSU\n"
+        "Usage: su [options] [argument...]\n\n"
+        "Options: \n"
+        "-c Pass command to invoked shell\n"
+        "-v Print version code and exit\n"
+        "-h Display this message and exit\n"
+    );
+    exit(status);
+
+}
+
+static bool ksuctl(int cmd, void * arg1, void * arg2) {
+
     int32_t result = 0;
-    prctl(0xdeadbeef, 0, 0, 0, &result);
-    system("/system/bin/sh");
+    prctl(0xDEADBEEF, cmd, arg1, arg2, & result);
+    return result == 0xDEADBEEF;
+
+}
+
+void elevate() {
+
+    // Talk to Daemon in Kernel Space
+    bool status = ksuctl(CMD_GRANT_ROOT, 0, NULL);
+
+    if (status) {
+        fprintf(stderr, "Permission denied\n");
+        exit(EXIT_FAILURE);
+    }
+
+}
+
+int getver() {
+
+    elevate();
+
+    int32_t version = -1;
+    if (ksuctl(CMD_GET_VERSION, & version, NULL)) {
+        return version;
+    }
+    return version;
+
+}
+
+void su_shell(char * envp[]) {
+
+    char * shell = getenv("SHELL");
+    int status;
+
+    elevate();
+
+    if (shell) {
+        status = execve(shell, NULL, envp);
+    } else {
+        status = execve("/system/bin/sh", NULL, envp);
+    }
+
+    exit(status);
+}
+
+void su_command(char * command[], int len) {
+
+    elevate();
+    int status;
+
+    if (len == 1) {
+        status = system(command[0]);
+    } else {
+        command[len] = NULL;
```

Comment:
Is this necessary? the args passed by `main` is already null-terminated(maybe you should check it)?

Reply:
Seems works without it, I will delete it.